daemon/logger/jsonfile/utils.go

```diff
@@ -0,0 +1,229 @@
+	watchTimeout := time.NewTimer(time.Second)
+	defer watchTimeout.Stop()
+
+	// handleError will watch the file if the err is io.EOF so that
+	// the loop can continue to read the file. Or just return the error.
+	handleError := func(err error) error {
+		if err != io.EOF {
+			return err
+		}
+
+		for {
+			watchTimeout.Reset(watchFileTimeout)
+
+			select {
+			case <-ctx.Done():
+				// caused by the caller and we should return done
+				return errDone
+			case e := <-fileWatcher.Events:
+				switch e.Op {
+				case fsnotify.Write:
+					decodeOneLine = unmarshaler(f)
+					return nil
+				case fsnotify.Remove:
+					// ideally, it's caused by removing the container.
+					return errDone
+				default:
+					logrus.Debug("unexpected file change during watching file %v: %v", f.Name(), e.Op)
+					return errDone
+				}
+			case newErr := <-fileWatcher.Errors:
+				// something wrong during the watching.
+				logrus.Debug("unexpected error during watching file %v: %v", f.Name(), newErr)
+				return err
+			case <-watchTimeout.C:
+				// FIXME: Since we hold the file handler in the process,
+				// the fsnotify cannot propagate the Remove event.
+				// This is workaround....
+				//
+				// More detail: https://github.com/fsnotify/fsnotify/issues/194
+				_, sErr := os.Stat(f.Name())
+				if sErr != nil {
+					if os.IsNotExist(sErr) {
+						return errDone
+					}
+					logrus.Debug("unexpected error during watching file %v: %v", f.Name(), sErr)
+					return errDone
+				}
+			}
+		}
+	}
+
+	// the dead loop to continue to read log
+	for {
+		msg, err := decodeOneLine()
+		if err != nil {
+			if err = handleError(err); err != nil {
+				if err == errDone {
+					return
+				}
+
+				watcher.Err <- err
+				return
+			}
+			continue
+		}
+
+		if !cfg.Since.IsZero() && msg.Timestamp.Before(cfg.Since) {
+			continue
+		}
+
+		if !cfg.Until.IsZero() && msg.Timestamp.After(cfg.Until) {
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case watcher.Msgs <- msg:
+		}
+	}
+}
+
+// watchFileChange will watch the change of file.
+func watchFileChange(filePath string) (*fsnotify.Watcher, error) {
+	fileWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := fileWatcher.Add(filePath); err != nil {
+		return nil, err
+	}
+	return fileWatcher, nil
+}
+
+// tailFile will read the log message until the io.EOF or limited by config.
+func tailFile(r io.Reader, cfg *logger.ReadConfig, unmarshaler newUnmarshalFunc, watcher *logger.LogWatcher) {
+	decodeOneLine := unmarshaler(r)
+
+	for {
+		msg, err := decodeOneLine()
+		if err != nil {
+			if err != io.EOF {
+				watcher.Err <- err
+			}
+			return
+		}
+
+		if !cfg.Since.IsZero() && msg.Timestamp.Before(cfg.Since) {
+			continue
+		}
+
+		if !cfg.Until.IsZero() && msg.Timestamp.After(cfg.Until) {
+			return
+		}
+
+		select {
+		case <-watcher.WatchClose():
+			return
+		case watcher.Msgs <- msg:
+		}
+	}
+}
+
+const (
+	blockSize = 1024
+	eol       = '\n'
```

Comment:
how about using `endOfLine` intead of `eol` which seems to be less readable. 😄

Reply:
OK. Will do